src/node_url_pattern.cc

```diff
@@ -0,0 +1,791 @@
+      if (!options) {
+        THROW_ERR_INVALID_ARG_TYPE(env, "options.ignoreCase must be a boolean");
+        return;
+      }
+    }
+  }
+
+  // Either url_pattern_init or input as a string must be provided.
+  CHECK_IMPLIES(init.has_value(), !input.has_value());
+
+  std::string_view base_url_view{};
+  if (base_url) base_url_view = {base_url->data(), base_url->size()};
+
+  ada::url_pattern_input arg0;
+  if (init.has_value()) {
+    arg0 = std::move(*init);
+  } else {
+    arg0 = std::move(*input);
+  }
+
+  auto url_pattern = parse_url_pattern<URLPatternRegexProvider>(
+      arg0,
+      base_url ? &base_url_view : nullptr,
+      options.has_value() ? &options.value() : nullptr);
+
+  if (!url_pattern) {
+    THROW_ERR_INVALID_URL_PATTERN(env, "Failed to constuct URLPattern");
+    return;
+  }
+
+  new URLPattern(env, args.This(), std::move(*url_pattern));
+}
+
+MaybeLocal<Value> URLPattern::URLPatternInit::ToJsObject(
+    Environment* env, const ada::url_pattern_init& init) {
+  auto isolate = env->isolate();
+  auto context = env->context();
+  auto result = Object::New(isolate);
+  if (init.protocol) {
+    Local<Value> protocol;
+    if (!ToV8Value(context, *init.protocol).ToLocal(&protocol)) {
+      return {};
+    }
+    USE(result->Set(context, env->protocol_string(), protocol));
+  }
+  if (init.username) {
+    Local<Value> username;
+    if (!ToV8Value(context, *init.username).ToLocal(&username)) {
+      return {};
+    }
+    USE(result->Set(context, env->username_string(), username));
+  }
+  if (init.password) {
+    Local<Value> password;
+    if (!ToV8Value(context, *init.password).ToLocal(&password)) {
+      return {};
+    }
+    USE(result->Set(context, env->password_string(), password));
+  }
+  if (init.hostname) {
+    Local<Value> hostname;
+    if (!ToV8Value(context, *init.hostname).ToLocal(&hostname)) {
+      return {};
+    }
+    USE(result->Set(context, env->hostname_string(), hostname));
+  }
+  if (init.port) {
+    Local<Value> port;
+    if (!ToV8Value(context, *init.port).ToLocal(&port)) {
+      return {};
+    }
+    USE(result->Set(context, env->port_string(), port));
+  }
+  if (init.pathname) {
+    Local<Value> pathname;
+    if (!ToV8Value(context, *init.pathname).ToLocal(&pathname)) {
+      return {};
+    }
+    USE(result->Set(context, env->pathname_string(), pathname));
+  }
+  if (init.search) {
+    Local<Value> search;
+    if (!ToV8Value(context, *init.search).ToLocal(&search)) {
+      return {};
+    }
+    USE(result->Set(context, env->search_string(), search));
+  }
+  if (init.hash) {
+    Local<Value> hash;
+    if (!ToV8Value(context, *init.hash).ToLocal(&hash)) {
+      return {};
+    }
+    USE(result->Set(context, env->hash_string(), hash));
+  }
+  if (init.base_url) {
+    Local<Value> base_url;
+    if (!ToV8Value(context, *init.base_url).ToLocal(&base_url)) {
+      return {};
+    }
+    USE(result->Set(context, env->base_url_string(), base_url));
```

Comment:
I don't understand why we do proper error checking for `ToV8Value()` but skip it for `Set()` 🙂

Reply:
Yep, these should be handled correctly. `USE()` is effectively just ignoring the possible error.

Reply:
You're both correct. I hate the endless if statements that V8 is forcing me... Is there a better pattern to avoid this much if statement, or 1 if statement with a lot bool expressions?

Reply:
> or 1 if statement with a lot bool expressions?

Yeah, I don't like it either but this is often the most convenient way to do this